### PR TITLE
Implement indirect calls.

### PR DESCRIPTION
**Pull request by ptersilie**
Companion PR: https://github.com/ykjit/ykllvm/pull/151

### CODE REVIEW

File: tests/c/indirect_call.newcg.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+__attribute__((noinline)) int foo(int i) {
+  return i + 3;
+}
+
+int main(int argc, char **argv) {
+  YkMT *mt = yk_mt_new(NULL);
+  yk_mt_hot_threshold_set(mt, 0);
+  YkLocation loc = yk_location_new();
+
+  int res = 9998;
```
Comment (vext01) [inline tests/c/indirect_call.newcg.c:45]: Does `res` play any role here? Kill it?

File: tests/c/indirect_call.newcg.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
+// ignore-if: test $YK_JIT_COMPILER != "yk" -o "$YKB_TRACER" = "swt"
+// Run-time:
+//   env-var: YKD_LOG_IR=-:aot,jit-pre-opt
+//   env-var: YKD_SERIALISE_COMPILATION=1
+//   env-var: YKD_LOG_JITSTATE=-
+//   stderr:
+//     jitstate: start-tracing
+//     foo 7
+//     jitstate: stop-tracing
+//     --- Begin aot ---
```
Comment (vext01) [inline tests/c/indirect_call.newcg.c:10]: Is the AOT match checking anything useful here?

File: tests/c/simple.c [hunk 3]
```diff
@@ -1,5 +1,5 @@
 // Run-time:
-//   env-var: YKD_LOG_IR=-:jit-pre-opt
+//   env-var: YKD_LOG_IR=-:aot,jit-pre-opt
```
Comment (vext01) [inline tests/c/simple.c:2]: Accidental change?

File: tests/c/indirect_call.newcg.c (hunk 1 above)
Comment (ptersilie) [reply tests/c/indirect_call.newcg.c:45]: It doesn't. Will do.

File: tests/c/simple.c (hunk 3 above)
Comment (ptersilie) [reply tests/c/simple.c:2]: Yes. Good spot.

File: tests/c/indirect_call.newcg.c (hunk 2 above)
Comment (ptersilie) [reply tests/c/indirect_call.newcg.c:10]: No. It was just copied over from the other test. I can remove it.

File: ykrt/src/compile/jitc_yk/aot_ir.rs [hunk 4]
```diff
@@ -713,6 +735,18 @@ impl AotIRDisplay for Instruction {
...
+                fty_idx: _,
+                callop,
+                args,
+            } => {
+                let args_s = args
+                    .iter()
+                    .map(|a| a.to_string(m))
+                    .collect::<Vec<_>>()
+                    .join(", ");
+                ret.push_str(&format!("call {}({})", callop.to_string(m), args_s,));
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/aot_ir.rs:748]: Did `cargo fmt` allow that trailing comma?

File: ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs [hunk 5]
```diff
@@ -463,6 +464,66 @@ impl<'a> X64CodeGen<'a> {
         self.emit_call(inst_idx, func_decl_idx, &args)
     }
 
+    /// Codegen a (non-varargs) call.
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:467]: I think the docstring needs updating?

File: ykrt/src/compile/jitc_yk/aot_ir.rs (hunk 4 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/aot_ir.rs:748]: Seems so.

File: ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs [hunk 6]
```diff
@@ -463,6 +464,66 @@ impl<'a> X64CodeGen<'a> {
...
+    fn cg_indirectcall(
+        &mut self,
+        inst_idx: InstIdx,
+        indirect_call_idx: &IndirectCallIdx,
+    ) -> Result<(), CompilationError> {
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:472]: Could any of `emit_call` be re-used (maybe with some adaption)?

File: ykrt/src/compile/jitc_yk/jit_ir.rs [hunk 7]
```diff
@@ -1237,6 +1285,71 @@ impl LookupGlobalInst {
         self.global_decl_idx
     }
 }
+///
+/// The operands for a [Inst::Call]
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/jit_ir.rs:1289]: Update the docstring to reference a `IndirectCall` instruction?

File: ykrt/src/compile/jitc_yk/jit_ir.rs [hunk 8]
```diff
@@ -1237,6 +1285,71 @@ impl LookupGlobalInst {
...
+///
+/// # Semantics
+///
+/// Perform a call to an external or AOT function.
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/jit_ir.rs:1293]: Perform an *indirect* call?

File: ykrt/src/compile/jitc_yk/trace_builder.rs [hunk 9]
```diff
@@ -516,6 +525,56 @@ impl<'a> TraceBuilder<'a> {
...
+        // Convert AOT args to JIT args.
+        let mut jit_args = Vec::new();
+        for arg in args {
+            jit_args.push(self.handle_operand(arg)?);
+        }
+
+        // While we can work out if this is a mappable call or not by loooking at the next block,
+        // this unfortunately doesn't tell us whether the call has been annotated with
+        // "yk_outline". For now we have to be conservative here and outline all indirect calls.
+        // One solution would be to stop including the AOR IR for functions annotated with
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/trace_builder.rs:549]: AOR -> AOT

File: ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs (hunk 6 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:472]: Yes. But I tried but gave up due to the different arguments each require.

**Comment by vext01** [comment]
Nice work. I have only minor comments.

File: ykrt/src/compile/jitc_yk/aot_ir.rs (hunk 4 above)
Comment (vext01) [reply ykrt/src/compile/jitc_yk/aot_ir.rs:748]: Fair enough. Remove the comma if you want.

File: ykrt/src/compile/jitc_yk/jit_ir.rs (hunk 7 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/jit_ir.rs]: [31e98ca](https://github.com/ykjit/yk/pull/1141/commits/31e98cac0bc2a5989282aa4fc0c0c2d0537230a5)

File: ykrt/src/compile/jitc_yk/jit_ir.rs [hunk 10]
```diff
@@ -189,6 +197,19 @@ impl Module {
         }
     }
 
+    pub(crate) fn push_indirectcall(
```
[the remark this answers is not in the document]
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/jit_ir.rs]: [6d5894c](https://github.com/ykjit/yk/pull/1141/commits/6d5894c9c5594b9700d9649f601d943159656c70)

File: ykrt/src/compile/jitc_yk/trace_builder.rs (hunk 9 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/trace_builder.rs]: [65b16f6](https://github.com/ykjit/yk/pull/1141/commits/65b16f6489b777779cb12d0527f5d780494b72aa)

File: ykrt/src/compile/jitc_yk/jit_ir.rs (hunk 8 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/jit_ir.rs]: [65b16f6](https://github.com/ykjit/yk/pull/1141/commits/65b16f6489b777779cb12d0527f5d780494b72aa)

File: tests/c/simple.c (hunk 3 above)
Comment (ptersilie) [reply tests/c/simple.c]: [98d9ebb](https://github.com/ykjit/yk/pull/1141/commits/98d9ebb9d0e73e6d360a5bc34d99747779e0dfcf)

File: tests/c/indirect_call.newcg.c (hunk 2 above)
Comment (ptersilie) [reply tests/c/indirect_call.newcg.c]: [9883a3a](https://github.com/ykjit/yk/pull/1141/commits/9883a3a605edc40a905df1fed200a5c473bc3fb6)

File: ykrt/src/compile/jitc_yk/aot_ir.rs (hunk 4 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/aot_ir.rs]: 21b461be499ecfc90c506161e0db230d24ef9699

**Comment by ptersilie** [comment]
Fixed comments. Note the additional commit [117137e](https://github.com/ykjit/yk/pull/1141/commits/117137e814caefa8ac386eeafd33c6108b8e878a) which renames `CallInst` to `DirectCallInst`

File: ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs (hunk 6 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:473]: [d6e6c4f](https://github.com/ykjit/yk/pull/1141/commits/d6e6c4fff6090c26c9f785c5cc385de523d3f39d)

Comment (vext01) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:473]: The opposite. We might be able to funnel indirect calls through `emit_call`?

Comment (vext01) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:473]: Oh wait. i think I see what you mean.

Comment (vext01) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:473]: Yeah, I think remove the comment:
```
// FIXME: This call can probably be inlined, as it's not used anywhere else.
```

I think we will re-use emit_call at some point (maybe).

Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:473]: I don't think we will tbh. But I'll remove the comment.

Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:472]: [00197c1](https://github.com/ykjit/yk/pull/1141/commits/00197c13d0e66d0f23b944c62a237497d5cb25a7)

**Comment by ptersilie** [comment]
Ok to squash?

**Comment by vext01** [comment]
Go ahead.

**Comment by ptersilie** [comment]
Squashed.

**Comment by ltratt** [comment]
cfmt

**Comment by vext01** [comment]
Please force push cfmt'd commits.

**Comment by ptersilie** [comment]
Done.

**Comment by ptersilie** [comment]
Maybe you should check this clippy change first since I had to disable a clippy feature to get it through.

**Comment by vext01** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed.

**Pull request merged**